File: gateway/ts/src/core/event.ts
import { authorize, type CallContext, caller } from "./context.js";
import {
  type Backpressure,
  type EventDescriptor,
  type EventSink,
  GatewayFailure,
  type Subscription,
  validateName,
} from "./protocol.js";

export interface LocalEventExport extends EventDescriptor {
  validate(filter: Uint8Array | undefined): void;
  matches(payload: Uint8Array, filter: Uint8Array | undefined): boolean;
}
export interface RemoteEventExport extends EventDescriptor {
  attach(context: CallContext, filter: Uint8Array | undefined, sink: EventSink): Promise<Subscription>;
}
export type EventExport = LocalEventExport | RemoteEventExport;
export class DeliveryQueue implements Subscription {
  private pending: Uint8Array[] = [];
  private draining = false;
  private active = true;
  constructor(
    public policy: Backpressure,
    private sink: EventSink,
  ) {}
  enqueue(payload: Uint8Array): void {
    if (!this.active || (this.policy === "drop" && this.pending.length)) return;
    if (this.policy === "coalesce") this.pending.length = 0;
    this.pending.push(payload.slice());
    if (this.draining) return;
    this.draining = true;
    queueMicrotask(() => {
      void this.drain();
    });
  }
  private async drain(): Promise<void> {
    while (this.active && this.pending.length) {
      const payload = this.pending.shift();
      if (!payload) break;
      try {
        await this.sink(payload);
      } catch {
        /* best-effort, no retries */
      }
    }
    this.draining = false;
  }
  clear(): void {
    this.pending.length = 0;
  }
  close(): void {
    this.active = false;
    this.clear();
  }
}
interface Entry {
  owner: object;
  name: string;
  export: EventExport;
}
interface Consumer {
  context: CallContext;
  active: boolean;
  generation: number;
  binding?: Subscription;
  caller: string;
  event: string;
  filter: Uint8Array | undefined;
  persistent: boolean;
  queue: DeliveryQueue;
}
export class Events {
  private entries = new Map<string, Entry>();
  private consumers = new Set<Consumer>();

  validateOwner(name: string, exports: readonly EventExport[]): void {
    const names = new Set<string>();
    for (const entry of exports) {
      validateName(entry.name);
      if (!["ordered", "coalesce", "drop"].includes(entry.policy)) {
        throw new GatewayFailure({ code: "INVALID_ARGUMENT", message: "invalid event policy" });
      }
      const existing = this.entries.get(entry.name);
      if (names.has(entry.name) || (existing && existing.name !== name)) {
        throw new GatewayFailure({ code: "CONFLICT", message: "event already registered" });
      }
      names.add(entry.name);
    }
  }
  install(owner: object, name: string, exports: readonly EventExport[]): void {
    for (const entry of exports) {
      this.entries.set(entry.name, { owner, name, export: { ...entry } });
      for (const consumer of this.consumers) {
        if (consumer.event === entry.name) {
          consumer.queue.policy = entry.policy;
          void this.activate(consumer).catch(() => {
            /* Invalid pending filters remain inactive. */
          });
        }
      }
    }
  }
  remove(owner: object): void {
    for (const [name, entry] of this.entries) {
      if (entry.owner !== owner) continue;
      this.entries.delete(name);
      for (const consumer of this.consumers) {
        if (consumer.event !== name) continue;
        consumer.generation++;
        if (consumer.binding) this.closeBinding(consumer.binding);
        consumer.binding = undefined;
        consumer.queue.clear();
        if (!consumer.persistent) this.close(consumer);
      }
    }
  }
  async subscribe(
    context: CallContext,
    event: string,
    filter: Uint8Array | undefined,
    sink: EventSink,
    persistent: boolean,
  ): Promise<Subscription> {
    authorize(context, event, true);
    validateName(event);
    const entry = this.entries.get(event);
    if (!entry && !persistent) throw new GatewayFailure({ code: "UNKNOWN_ROUTE", message: "event not exported" });
    if (entry && "validate" in entry.export) entry.export.validate(filter);
    const consumer: Consumer = {
      context,
      active: true,
      generation: 0,
      caller: caller(context),
      event,
      filter: filter?.slice(),
      persistent,
      queue: new DeliveryQueue(entry?.export.policy ?? "drop", sink),
    };
    this.consumers.add(consumer);
    try {
      await this.activate(consumer);
    } catch (error) {
      if (
        !(
          persistent &&
          error instanceof GatewayFailure &&
          ["UNKNOWN_ROUTE", "OWNER_UNAVAILABLE"].includes(error.detail.code)
        )
      ) {
        this.close(consumer);
        throw error;
      }
    }
    return { close: () => this.close(consumer) };
  }
  private closeBinding(binding: Subscription): void {
    try {
      binding.close();
    } catch {
      /* A disconnected endpoint cannot prevent local cleanup. */
    }
  }
  private close(consumer: Consumer): void {
    consumer.active = false;
    consumer.generation++;
    if (consumer.binding) this.closeBinding(consumer.binding);
    consumer.binding = undefined;
    consumer.queue.close();
    this.consumers.delete(consumer);
  }
  private async activate(consumer: Consumer): Promise<void> {
    const entry = this.entries.get(consumer.event);
    if (!entry || !consumer.active) return;
    const generation = ++consumer.generation;
    if ("validate" in entry.export) {
      entry.export.validate(consumer.filter);
      return;
    }
    let binding: Subscription;
    try {
      binding = await entry.export.attach(consumer.context, consumer.filter, (payload) => {
        if (consumer.active && consumer.generation === generation && this.entries.get(consumer.event) === entry) {
          consumer.queue.enqueue(payload);
        }
      });
    } catch (error) {
      if (!consumer.active || consumer.generation !== generation || this.entries.get(consumer.event) !== entry) return;
      throw error;
    }
    if (!consumer.active || consumer.generation !== generation || this.entries.get(consumer.event) !== entry) {
      this.closeBinding(binding);
      return;
    }
    consumer.binding = binding;
  }
  publish(owner: object, event: string, payload: Uint8Array): void {
    const entry = this.entries.get(event);
    if (!entry) throw new GatewayFailure({ code: "UNKNOWN_ROUTE", message: "event not exported" });
    if (entry.owner !== owner) throw new GatewayFailure({ code: "OWNER_UNAVAILABLE", message: "stale event source" });
    if (!("matches" in entry.export)) {
      throw new GatewayFailure({ code: "INVALID_ARGUMENT", message: "remote events use subscription delivery" });
    }
    for (const consumer of this.consumers) {
      if (consumer.event !== event) continue;
      try {
        entry.export.validate(consumer.filter);
        if (entry.export.matches(payload, consumer.filter)) consumer.queue.enqueue(payload);
      } catch {
        /* Invalid pending filters never receive events after owner registration. */
      }
    }
  }
  cleanupCaller(id: string): void {
    for (const consumer of this.consumers) {
      if (consumer.caller === id) this.close(consumer);
    }
  }
}

File: gateway/rust/src/event.rs
//! Explicit event exports and per-subscription delivery queues, extracted from
//! xw-tauri/event.rs. Unavailable sources retain subscription intent, not history.
use std::collections::{HashMap, HashSet, VecDeque};
use std::future::Future;
use std::pin::Pin;
use std::sync::{Arc, Mutex};

use prost::Message;
use serde::{Deserialize, Serialize};

use crate::invoke::{Owner, XwInvokeRegistry};
use crate::protocol::*;

type Validator = Arc<dyn Fn(Option<&[u8]>) -> Result<(), GatewayError> + Send + Sync>;
type Matcher = Arc<dyn Fn(&[u8], Option<&[u8]>) -> bool + Send + Sync>;
type Sink = Arc<dyn Fn(Vec<u8>) -> Pin<Box<dyn Future<Output = ()> + Send>> + Send + Sync>;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum EventBackpressure {
    Ordered,
    Coalesce,
    Drop,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EventDescriptor {
    pub name: String,
    pub policy: EventBackpressure,
}

#[derive(Clone)]
pub struct EventExportRegistration {
    pub name: String,
    pub policy: EventBackpressure,
    validate: Validator,
    matches: Matcher,
}

impl EventExportRegistration {
    pub fn typed<Payload, Filter, F>(policy: EventBackpressure, matcher: F) -> Self
    where
        Payload: Message + Default + prost::Name,
        Filter: Message + Default,
        F: Fn(&Payload, Option<&Filter>) -> bool + Send + Sync + 'static,
    {
        Self {
            name: Payload::full_name(),
            policy,
            validate: Arc::new(|filter| match filter {
                Some(bytes) => Filter::decode(bytes)
                    .map(|_| ())
                    .map_err(|_| GatewayError::new(ErrorCode::InvalidArgument, "invalid event filter")),
                None => Ok(()),
            }),
            matches: Arc::new(move |payload, filter| {
                let Ok(payload) = Payload::decode(payload) else {
                    return false;
                };
                let filter = match filter {
                    Some(bytes) => match Filter::decode(bytes) {
                        Ok(value) => Some(value),
                        Err(_) => return false,
                    },
                    None => None,
                };
                matcher(&payload, filter.as_ref())
            }),
        }
    }
}

struct QueueState {
    policy: EventBackpressure,
    pending: VecDeque<Vec<u8>>,
    draining: bool,
    active: bool,
}

struct DeliveryQueue {
    state: Mutex<QueueState>,
    sink: Sink,
}

impl DeliveryQueue {
    fn enqueue(self: &Arc<Self>, payload: Vec<u8>) {
        let mut state = self.state.lock().unwrap();
        if !state.active {
            return;
        }
        match state.policy {
            EventBackpressure::Drop if !state.pending.is_empty() => return,
            EventBackpressure::Coalesce => state.pending.clear(),
            _ => {}
        }
        state.pending.push_back(payload);
        if state.draining {
            return;
        }
        state.draining = true;
        drop(state);
        let queue = self.clone();
        tokio::spawn(async move {
            loop {
                let payload = {
                    let mut state = queue.state.lock().unwrap();
                    match state.pending.pop_front() {
                        Some(payload) if state.active => payload,
                        _ => {
                            state.draining = false;
                            return;
                        }
                    }
                };
                // Isolate sink panics so one subscriber cannot strand another queue.
                let sink = queue.sink.clone();
                let _ = tokio::spawn(async move { sink(payload).await }).await;
            }
        });
    }
    fn clear(&self) {
        self.state.lock().unwrap().pending.clear();
    }
    fn close(&self) {
        let mut state = self.state.lock().unwrap();
        state.active = false;
        state.pending.clear();
    }
}

struct Subscription {
    name: String,
    filter: Option<Vec<u8>>,
    caller: String,
    persistent: bool,
    queue: Arc<DeliveryQueue>,
}

#[derive(Default)]
pub(crate) struct EventState {
    entries: HashMap<String, (Owner, EventExportRegistration)>,
    subscriptions: HashMap<u64, Subscription>,
    next_id: u64,
}

impl EventState {
    pub(crate) fn manifest(&self, owner: &Owner) -> Vec<EventDescriptor> {
        let mut events: Vec<_> = self
            .entries
            .values()
            .filter(|(current, _)| current.same_instance(owner))
            .map(|(_, export)| EventDescriptor {
                name: export.name.clone(),
                policy: export.policy,
            })
            .collect();
        events.sort_by(|a, b| a.name.cmp(&b.name));
        events
    }
    pub(crate) fn validate_owner(&self, name: &str, exports: &[EventExportRegistration]) -> Result<(), GatewayError> {
        let mut names = HashSet::new();
        for export in exports {
            validate_name(&export.name)?;
            if !names.insert(&export.name)
                || self
                    .entries
                    .get(&export.name)
                    .is_some_and(|(owner, _)| owner.name != name)
            {
                return Err(GatewayError::new(ErrorCode::Conflict, "event already registered"));
            }
        }
        Ok(())
    }
    pub(crate) fn install_owner(&mut self, owner: &Owner, exports: Vec<EventExportRegistration>) {
        for export in exports {
            for subscription in self.subscriptions.values().filter(|s| s.name == export.name) {
                subscription.queue.state.lock().unwrap().policy = export.policy;
            }
            self.entries.insert(export.name.clone(), (owner.clone(), export));
        }
    }
    pub(crate) fn remove_owner(&mut self, owner: &Owner) {
        let names: HashSet<_> = self
            .entries
            .iter()
            .filter(|(_, (o, _))| o.same_instance(owner))
            .map(|(name, _)| name.clone())
            .collect();
        self.entries.retain(|_, (o, _)| !o.same_instance(owner));
        self.subscriptions.retain(|_, subscription| {
            if !names.contains(&subscription.name) {
                return true;
            }
            subscription.queue.clear();
            if subscription.persistent {
                true
            } else {
                subscription.queue.close();
                false
            }
        });
    }
}

/// Opaque subscription handle. Dropping it unsubscribes; close is idempotent.
pub struct SubscriptionHandle {
    registry: Arc<XwInvokeRegistry>,
    id: u64,
}
impl SubscriptionHandle {
    pub fn close(&self) {
        if let Some(subscription) = self
            .registry
            .state
            .lock()
            .unwrap()
            .events
            .subscriptions
            .remove(&self.id)
        {
            subscription.queue.close();
        }
    }
}
impl Drop for SubscriptionHandle {
    fn drop(&mut self) {
        self.close();
    }
}

impl XwInvokeRegistry {
    pub fn snapshot_events(&self) -> Vec<(String, EventBackpressure)> {
        let state = self.state.lock().unwrap();
        let mut entries: Vec<_> = state
            .events
            .entries
            .iter()
            .map(|(name, (_, e))| (name.clone(), e.policy))
            .collect();
        entries.sort_by(|a, b| a.0.cmp(&b.0));
        entries
    }

    pub fn subscribe<F, Fut>(
        self: &Arc<Self>,
        context: &CallContext,
        name: &str,
        filter: Option<Vec<u8>>,
        persistent: bool,
        sink: F,
    ) -> Result<SubscriptionHandle, GatewayError>
    where
        F: Fn(Vec<u8>) -> Fut + Send + Sync + 'static,
        Fut: Future<Output = ()> + Send + 'static,
    {
        context.authorize(name, true)?;
        validate_name(name)?;
        let mut state = self.state.lock().unwrap();
        let policy = if let Some((_, export)) = state.events.entries.get(name) {
            (export.validate)(filter.as_deref())?;
            export.policy
        } else if persistent {
            EventBackpressure::Drop
        } else {
            return Err(GatewayError::new(ErrorCode::UnknownRoute, "event not exported"));
        };
        state.events.next_id += 1;
        let id = state.events.next_id;
        state.events.subscriptions.insert(
            id,
            Subscription {
                name: name.into(),
                filter,
                caller: context.caller().into(),
                persistent,
                queue: Arc::new(DeliveryQueue {
                    state: Mutex::new(QueueState {
                        policy,
                        pending: VecDeque::new(),
                        draining: false,
                        active: true,
                    }),
                    sink: Arc::new(move |payload| Box::pin(sink(payload))),
                }),
            },
        );
        Ok(SubscriptionHandle {
            registry: self.clone(),
            id,
        })
    }

    pub fn cleanup_caller(&self, caller: &str) {
        self.remote_events.cleanup_caller(caller);
        self.state
            .lock()
            .unwrap()
            .events
            .subscriptions
            .retain(|_, subscription| {
                if subscription.caller != caller {
                    return true;
                }
                subscription.queue.close();
                false
            });
    }

    pub fn publish(&self, owner: &Owner, name: &str, payload: Vec<u8>) -> Result<(), GatewayError> {
        let state = self.state.lock().unwrap();
        let (current, export) = state
            .events
            .entries
            .get(name)
            .ok_or_else(|| GatewayError::new(ErrorCode::UnknownRoute, "event not exported"))?;
        if !current.same_instance(owner) || owner.is_closed() {
            return Err(GatewayError::new(ErrorCode::OwnerUnavailable, "stale event source"));
        }
        for subscription in state.events.subscriptions.values().filter(|s| s.name == name) {
            if (export.validate)(subscription.filter.as_deref()).is_ok()
                && (export.matches)(&payload, subscription.filter.as_deref())
            {
                subscription.queue.enqueue(payload.clone());
            }
        }
        Ok(())
    }
}

/// The transport supplies a connection-bound subscription. Closing it must be
/// idempotent and target only that connection instance.
pub struct RemoteBinding {
    pub policy: EventBackpressure,
    pub close: Arc<dyn Fn() + Send + Sync>,
}
impl Drop for RemoteBinding {
    fn drop(&mut self) {
        (self.close)();
    }
}

pub type EventSink = Arc<dyn Fn(Vec<u8>) + Send + Sync>;
type SubscribeFuture = Pin<Box<dyn Future<Output = Result<RemoteBinding, GatewayError>> + Send>>;
pub type EventTransport = Arc<dyn Fn(CallContext, String, Option<Vec<u8>>, EventSink) -> SubscribeFuture + Send + Sync>;

struct RemoteConsumer {
    context: CallContext,
    name: String,
    filter: Option<Vec<u8>>,
    persistent: bool,
    queue: Arc<DeliveryQueue>,
    binding: Option<RemoteBinding>,
    generation: u64,
}
#[derive(Default)]
struct RemoteState {
    next_id: u64,
    transport: Option<EventTransport>,
    consumers: HashMap<u64, RemoteConsumer>,
}

/// Consumer-side adapter. Replacing a transport invalidates old delivery closures
/// before replaying subscription intent; no event bytes are buffered offline.
#[derive(Clone, Default)]
pub struct RemoteEvents(Arc<Mutex<RemoteState>>);

pub struct RemoteSubscription {
    events: RemoteEvents,
    id: u64,
}
impl RemoteSubscription {
    pub fn close(&self) {
        // Drop the transport binding outside our lock (it can re-enter the registry).
        let removed = self.events.0.lock().unwrap().consumers.remove(&self.id);
        if let Some(consumer) = &removed {
            consumer.queue.close();
        }
        drop(removed);
    }
}
impl Drop for RemoteSubscription {
    fn drop(&mut self) {
        self.close();
    }
}

impl RemoteEvents {
    /// Permanently release transport and all subscription intents at endpoint shutdown.
    pub fn close(&self) {
        let (transport, consumers) = {
            let mut state = self.0.lock().unwrap();
            (state.transport.take(), std::mem::take(&mut state.consumers))
        };
        for consumer in consumers.values() {
            consumer.queue.close();
        }
        drop(consumers);
        drop(transport);
    }
    pub async fn set_transport(&self, transport: Option<EventTransport>) {
        let (bindings, removed, ids) = {
            let mut state = self.0.lock().unwrap();
            state.transport = transport;
            let mut bindings = Vec::new();
            let mut removed = Vec::new();
            for consumer in state.consumers.values_mut() {
                consumer.generation += 1;
                consumer.queue.clear();
                if let Some(binding) = consumer.binding.take() {
                    bindings.push(binding);
                }
            }
            let expired: Vec<_> = state
                .consumers
                .iter()
                .filter(|(_, c)| !c.persistent)
                .map(|(id, _)| *id)
                .collect();
            for id in expired {
                if let Some(consumer) = state.consumers.remove(&id) {
                    consumer.queue.close();
                    removed.push(consumer);
                }
            }
            (bindings, removed, state.consumers.keys().copied().collect::<Vec<_>>())
        };
        drop(bindings);
        drop(removed);
        for id in ids {
            let _ = self.activate(id).await;
        }
    }

    pub async fn subscribe<F, Fut>(
        &self,
        context: CallContext,
        name: &str,
        filter: Option<Vec<u8>>,
        persistent: bool,
        sink: F,
    ) -> Result<RemoteSubscription, GatewayError>
    where
        F: Fn(Vec<u8>) -> Fut + Send + Sync + 'static,
        Fut: Future<Output = ()> + Send + 'static,
    {
        context.authorize(name, true)?;
        validate_name(name)?;
        let id = {
            let mut state = self.0.lock().unwrap();
            state.next_id += 1;
            let id = state.next_id;
            state.consumers.insert(
                id,
                RemoteConsumer {
                    context,
                    name: name.into(),
                    filter,
                    persistent,
                    binding: None,
                    generation: 0,
                    queue: Arc::new(DeliveryQueue {
                        state: Mutex::new(QueueState {
                            policy: EventBackpressure::Drop,
                            pending: VecDeque::new(),
                            draining: false,
                            active: true,
                        }),
                        sink: Arc::new(move |payload| Box::pin(sink(payload))),
                    }),
                },
            );
            id
        };
        let handle = RemoteSubscription {
            events: self.clone(),
            id,
        };
        if let Err(error) = self.activate(id).await {
            if !persistent || !matches!(error.code, ErrorCode::UnknownRoute | ErrorCode::OwnerUnavailable) {
                handle.close();
                return Err(error);
            }
        }
        Ok(handle)
    }

    async fn activate(&self, id: u64) -> Result<(), GatewayError> {
        let (transport, context, name, filter, generation) = {
            let mut state = self.0.lock().unwrap();
            let transport = state
                .transport
                .clone()
                .ok_or_else(|| GatewayError::new(ErrorCode::OwnerUnavailable, "event transport unavailable"))?;
            let Some(consumer) = state.consumers.get_mut(&id) else {
                return Ok(());
            };
            consumer.generation += 1;
            (
                transport,
                consumer.context.clone(),
                consumer.name.clone(),
                consumer.filter.clone(),
                consumer.generation,
            )
        };
        let weak = Arc::downgrade(&self.0);
        let sink: EventSink = Arc::new(move |payload| {
            let Some(state) = weak.upgrade() else {
                return;
            };
            let state = state.lock().unwrap();
            if let Some(consumer) = state.consumers.get(&id) {
                // Before attach completes there is no acknowledged policy/lease.
                if consumer.generation == generation && consumer.binding.is_some() {
                    consumer.queue.enqueue(payload);
                }
            }
        });
        let result = transport(context, name, filter, sink).await;
        let mut state = self.0.lock().unwrap();
        if let Some(consumer) = state.consumers.get_mut(&id) {
            if consumer.generation == generation {
                let binding = result?;
                consumer.queue.state.lock().unwrap().policy = binding.policy;
                consumer.binding = Some(binding);
                return Ok(());
            }
        }
        drop(state);
        drop(result);
        Ok(())
    }

    pub fn cleanup_caller(&self, caller: &str) {
        let ids: Vec<_> = self
            .0
            .lock()
            .unwrap()
            .consumers
            .iter()
            .filter(|(_, c)| c.context.caller() == caller)
            .map(|(id, _)| *id)
            .collect();
        for id in ids {
            RemoteSubscription {
                events: self.clone(),
                id,
            }
            .close();
        }
    }
}

pub enum EventSubscription {
    Local(SubscriptionHandle),
    Remote(RemoteSubscription),
}
impl EventSubscription {
    pub fn close(&self) {
        match self {
            Self::Local(handle) => handle.close(),
            Self::Remote(handle) => handle.close(),
        }
    }
}

impl crate::invoke::Client {
    /// Local first. The injected host transport owns remote subscription replay
    /// when an event owner appears; transport replacement handles reconnection.
    pub async fn subscribe<F, Fut>(
        &self,
        name: &str,
        filter: Option<Vec<u8>>,
        persistent: bool,
        sink: F,
    ) -> Result<EventSubscription, GatewayError>
    where
        F: Fn(Vec<u8>) -> Fut + Send + Sync + 'static,
        Fut: Future<Output = ()> + Send + 'static,
    {
        self.context().authorize(name, true)?;
        let sink = Arc::new(sink);
        let local_sink = sink.clone();
        match self
            .registry()
            .subscribe(self.context(), name, filter.clone(), false, move |payload| {
                local_sink(payload)
            }) {
            Ok(handle) => Ok(EventSubscription::Local(handle)),
            Err(error) if error.code == ErrorCode::UnknownRoute => self
                .registry()
                .remote_events
                .subscribe(self.context().clone(), name, filter, persistent, move |payload| {
                    sink(payload)
                })
                .await
                .map(EventSubscription::Remote),
            Err(error) => Err(error),
        }
    }
}
